drift/core/beamtransfer.py

```diff
@@ -831,190 +831,264 @@ def _generate_svdfiles(self, regen=False, skip_svd_inv=False):
+                if nmodes == 0:
+                    continue
 
-                    # Final products
-                    ut = ut3
-                    sig = s3[:nmodes]
-                    beam = np.dot(ut3, bfr)
+                # Final products
+                ut = ut3
+                sig = s3[:nmodes]
+                beam = np.dot(ut3, bfr)
 
-                    # Save out the evecs (for transforming from the telescope frame into the SVD basis)
-                    dset_ut[fi, :nmodes] = ut * noisew[np.newaxis, :]
+                # Save out the evecs (for transforming from the telescope frame into the SVD basis)
+                dset_ut[fi, :nmodes] = ut * noisew[np.newaxis, :]
 
-                    # Save out the modified beam matrix (for mapping from the sky into the SVD basis)
-                    dset_bsvd[fi, :nmodes] = beam.reshape(
-                        nmodes, self.telescope.num_pol_sky, self.telescope.lmax + 1
-                    )
+                # Save out the modified beam matrix (for mapping from the sky into the SVD basis)
+                dset_bsvd[fi, :nmodes] = beam.reshape(
+                    nmodes, self.telescope.num_pol_sky, self.telescope.lmax + 1
+                )
 
-                    if not skip_svd_inv:
+                if not skip_svd_inv:
+                    # Find the pseudo-inverse of the beam matrix and save to disk.
+                    # First try la.pinv, which uses a least-squares solver.
+                    try:
                         ibeam = la.pinv(beam)
-                        # Find the pseudo-inverse of the beam matrix and save to disk.
-                        dset_ibsvd[fi, :, :, :nmodes] = ibeam.reshape(
-                            self.telescope.num_pol_sky, self.telescope.lmax + 1, nmodes
+                    except la.LinAlgError as e:
+                        # If la.pinv fails, try la.pinv2, which is SVD-based and
+                        # more likely to succeed. If successful, add file attribute
+                        # indicating pinv2 was used for this frequency.
+                        print(
+                            "***pinv failure: m = %d, fi = %d. Trying pinv2..."
+                            % (mi, fi)
                         )
+                        try:
+                            ibeam = la.pinv2(beam)
+                            if "inv_bsvd_from_pinv2" not in fs.attrs.keys():
+                                fs.attrs["inv_bsvd_from_pinv2"] = [fi]
+                            else:
+                                bad_freqs = fs.attrs["inv_bsvd_from_pinv2"]
+                                fs.attrs["inv_bsvd_from_pinv2"] = bad_freqs.append(fi)
+                        except:
+                            # If pinv2 fails, close SVD and beam files gracefully,
+                            # then print error message
+                            fs.close()
+                            fm.close()
```

Comment:
At the risk of a few more levels of indenting, you'd be better off just using a context manager for this.

Reply:
Done, with new commits from today